content/blog/snapshots.md

```diff
@@ -0,0 +1,75 @@
+date = 2023-11-20T00:09:00+02:00
+draft = false
+template = "blog/page.html"
+
+[taxonomies]
+authors = ["clementd"]
+
+[extra]
+lead = "What are biscuit snapshots, how do they work and how can they be useful"
++++
+
+One of the defining features in biscuit is the common language for authorization policies. Along with the cryptographic constructs used in tokens, it is what allows offline attuenation.
+
+A common language for policies means that we can have a standardized serialization format, which in turn means it can be embedded in a token, which gives us offline attenuation. Neat!
+
+It turns out that being able to serialize authorization policies gives us another benefit: it is possible to take a snapshot of an authorization process and save it for later. That's what we'll talk about today.
+
+## Authorizer snapshots
+
+In `biscuit-rust` and most biscuit libraries, the authorization process is carried out through an `Authorizer` value.
+
+An `Authorizer` is created from a biscuit token, along with facts, rules, checks, and policies added by the authorizing party. 
+
+Once all this has been provided, the `Authorizer` runs datalog evaluation (it repeatedly generates new datalog facts from rules unless no new facts can be generated). Once this is done, checks and policies are evaluated and are used to compute the authorization result (all checks have to pass, and the first policy to match must be an `allow` policy). The `Authorizer` makes sure these two steps are carried out in a timely fashion by aborting after a specified timeout, if too many facts are generated, or after a specific amount of iterations. This is crucial to make sure authorization does not become a DoS target.
+
+The good news is that an `Authorizer` only contains serializable data, and as such can be stored.
+
+```rust
+let mut authorizer = authorizer!(
+  r#"time({now});
+  resource("/file1.txt");
+  check if user($user);
+  allow if right("/file1.txt", read);
+  "#,
+  now = SystemTime::now(),
+);
+authorizer.add_token(biscuit);
+let result = authorizer.authorize();
+println!("{}", authorizer.snapshot().to_base64_snapshot())
+```
+
+This will give you something like:
+
+```
+ CgkI6AcQZBjAhD0Q2YkBGvMBCAQSCi9maWxlMS50eHQSBDEyMzQiRBADGgkKBwgKEgMYgQgaDQoLCAQSAxiACBICGAAqJgokCgIIGxIGCAUSAggFGhYKBAoCCAUKCAoGIIDEpKsGCgQaAggAKjUQAxoJCgcIAhIDGIAIGggKBggDEgIYABoMCgoIBRIGILCX3aoGKg4KDAoCCBsSBggKEgIICjIVChEKAggbEgsIBBIDGIAIEgIYABAAOicKAgoAEggKBggDEgIYABIJCgcIAhIDGIAIEgwKCggFEgYgsJfdqgY6HgoCEAASDQoLCAQSAxiACBICGAASCQoHCAoSAxiBCEAA 
+```
+
+Once you have that, you can inspect it with the CLI:
+
+```
+echo "CgkI6AcQZBjAhD0Q2YkBGvMBCAQSCi9maWxlMS50eHQSBDEyMzQiRBADGgkKBwgKEgMYgQgaDQoLCAQSAxiACBICGAAqJgokCgIIGxIGCAUSAggFGhYKBAoCCAUKCAoGIIDEpKsGCgQaAggAKjUQAxoJCgcIAhIDGIAIGggKBggDEgIYABoMCgoIBRIGILCX3aoGKg4KDAoCCBsSBggKEgIICjIVChEKAggbEgsIBBIDGIAIEgIYABAAOicKAgoAEggKBggDEgIYABIJCgcIAhIDGIAIEgwKCggFEgYgsJfdqgY6HgoCEAASDQoLCAQSAxiACBICGAASCQoHCAoSAxiBCEAA" | biscuit inspect-snapshot -
+```
```

Comment:
May be you can show the output of inspect-snapshot command.

Reply:
the output will be displayed right after, in the web component, so i imagined that would be too much to display it twice.